### PR TITLE
Added validations for save names

### LINKED ISSUE #2862: Save false success when saving with special symbols like `/`
image

image

```
Creating a save with name: {}[]|\/.thrivesave

ERROR: store_buffer: File must be opened before use.
   At: core/bind/core_bind.cpp:2265.
(The above error repeated for over a hundred times)

save finished, success: True message: Saving succeeded elapsed: 00:00:00.6036241
Jukebox: starting track: res://assets/sounds/soundeffects/microbe-ambience2.ogg position: 0

```

### PR DESCRIPTION
**Pull request by 84634E1A607A**
Fixes #2862 

**Progress Checklist**

Note: before starting this checklist the PR should be marked as non-draft.

- [x] PR author has checked that this PR works as intended and doesn't
      break existing features:
      https://wiki.revolutionarygamesstudio.com/wiki/Testing_Checklist
      (this is important as to not waste the time of Thrive team
      members reviewing this PR)
- [x] Initial code review passed (this and further items should not be checked by the PR author)
- [x] Functionality is confirmed working by another person (see above checklist link)
- [x] Final code review is passed and code conforms to the 
      [styleguide](https://github.com/Revolutionary-Games/Thrive/blob/master/doc/style_guide.md).

Before merging all CI jobs should finish on this PR without errors, if
there are automatically detected style issues they should be fixed by
the PR author.

### CODE REVIEW

**Comment by hhyyrylainen** [comment]
We are actually no longer in release freeze, forgot to turn off the bot immediately. It's off now.

**Comment by 84634E1A607A** [comment]
Review welcome.

**Review by Kasterisk — CHANGES_REQUESTED** [review]
I can still save with invalid name by pressing enter on the save name box.

**Comment by 84634E1A607A** [comment]
Done

**Comment by RacerBG** [comment]
I can still save with name like this: ~!@#$.

I'm doing the following steps **but something tells me that probably I'm messing up somewhere**:
1. git clone https://github.com/Revolutionary-Games/Thrive
2. cd Thrive
3. git fetch origin master:2862_alphanumeric_name
4. git checkout 2862_alphanumeric_name
5. Then I'm starting Godot.

**Comment by 84634E1A607A** [comment]
> ~!@#$

No banned characters here

**Comment by RacerBG** [comment]
> > ~!@#$
> 
> No banned characters here

I saw the code, you are right but still with them I can't save with "Enter" but the "Save" buttons works.

**Comment by 84634E1A607A** [comment]
That's pretty strange because these uses exactly the same logic. 
Will look into that later.

File: simulation_parameters/Constants.cs [hunk 1]
```diff
@@ -657,6 +658,8 @@ public static class Constants
     /// </summary>
     public const string SPECIES_NAME_REGEX = "^(?<genus>[^ ]+) (?<epithet>[^ ]+)$";
 
+    public const string SPECIES_NAME_DICT = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ1234567890 ";
+
```
Comment (FuJa0815) [inline simulation_parameters/Constants.cs:662]:
```suggestion
    public const string SPECIES_NAME_REGEX = "^(?<genus>[a-zA-Z0-9]+) (?<epithet>[a-zA-Z0-9]+)$";
```
you can merge these two checks into one

File: simulation_parameters/Constants.cs [hunk 2]
```diff
@@ -603,6 +603,7 @@ public static class Constants
     /// </summary>
     public const int MAX_JSON_ERROR_LENGTH_FOR_CONSOLE = 20000;
 
+    public const string SAVE_NAME_BLACKLIST = "<>:\"/\\|?*\0";
```
Comment (FuJa0815) [inline simulation_parameters/Constants.cs:606]: use `System.IO.Path.GetInvalidFileNameChars()`

File: src/saving/NewSaveMenu.cs [hunk 3]
```diff
@@ -132,10 +137,29 @@ private void OnSelectedChanged()
...
+        {
+            saveNameBox.Set("custom_colors/font_color", new Color(1.0f, 0.3f, 0.3f));
+            saveButton.Disabled = true;
+        }
+    }
 
-        SaveButtonPressed();
+    private void OnSaveNameTextEntered(string newName)
+    {
+        if (newName.Length != 0 && !newName.Any(Constants.SAVE_NAME_BLACKLIST.Contains))
```
Comment (FuJa0815) [inline src/saving/NewSaveMenu.cs:156]: make this check a private method.

Comment (FuJa0815) [reply src/saving/NewSaveMenu.cs:156]: And you can use `string.IsNullOrWhitespace()`

**Review by FuJa0815 — DISMISSED** [review]

File: simulation_parameters/Constants.cs (hunk 2 above)
Comment (hhyyrylainen) [reply simulation_parameters/Constants.cs:606]: Can't do that https://github.com/Revolutionary-Games/Thrive/issues/2862#issuecomment-980558005

**Comment by 84634E1A607A** [comment]
> I can still save with name like this by pressing the "Save" button: ~!@#$.
> 
> Pressing enter makes the text red and a pop up text is shown.

Cannot reproduce. Save succeeded by both pressing enter and click save button on my end.

**Comment by RacerBG** [comment]
> > I can still save with name like this by pressing the "Save" button: ~!@#$.
> > Pressing enter makes the text red and a pop up text is shown.
> 
> Cannot reproduce. Save succeeded by both pressing enter and click save button on my end.

Everything is fine after the latest changes. :+1:

File: src/saving/Save.cs [hunk 4]
```diff
@@ -231,7 +231,11 @@ public void DestroyGameStates()
     private static void WriteDataToSaveFile(string target, string justInfo, string serialized, string tempScreenshot)
     {
         using var file = new File();
-        file.Open(target, File.ModeFlags.Write);
+        if (file.Open(target, File.ModeFlags.Write) != Error.Ok)
+        {
+            GD.PrintErr("Cannot open file to write.");
```
Comment (hhyyrylainen) [inline src/saving/Save.cs:236]:
```suggestion
            GD.PrintErr("Cannot open file for writing: ", target);
```

File: src/saving/Save.cs [hunk 5]
```diff
@@ -231,7 +231,11 @@ public void DestroyGameStates()
...
+            throw new Exception("Cannot open: " + target);
```
Comment (hhyyrylainen) [inline src/saving/Save.cs:237]: Maybe IO.Error should be thrown instead to be more specific what's the problem?

File: simulation_parameters/Constants.cs (hunk 2 above)
Comment (hhyyrylainen) [inline simulation_parameters/Constants.cs:606]:
```suggestion
    public const string FILE_NAME_DISALLOWED_CHARACTERS = "<>:\"/\\|?*\0";
```

File: src/saving/NewSaveMenu.cs [hunk 6]
```diff
@@ -132,10 +137,29 @@ private void OnSelectedChanged()
         usingSelectedSaveName = true;
     }
 
-    private void OnSaveNameTextEntered(string newName)
+    private void OnSaveNameTextChanged(string newName)
     {
-        _ = newName;
+        if (!string.IsNullOrWhiteSpace(newName) && !newName.Any(Constants.SAVE_NAME_BLACKLIST.Contains))
```
Comment (hhyyrylainen) [inline src/saving/NewSaveMenu.cs:142]: I'd make this into a one liner helper method as this is a pretty long condition that needs to be kept in sync in two different methods, so having one helper method will guard against someone accidentally editing one place and not the other.

**Review by hhyyrylainen — CHANGES_REQUESTED** [review]
Looks mostly fine, I have a few code comments, after those are solved I'm fine with this being merged as other people have tested this enough, I think.

**Review by hhyyrylainen — APPROVED** [review]
I'll assume the functionality still works correctly now that the code comments are resolved.

**Pull request merged**